### ISSUE TITLE
[GR-32778] Keep  --enable-all-security-services

### ISSUE THREAD

**Issue by rmannibucau** [oracle/graal#3616]
Hi all,

not sure it is the best way to report it but read with last release annoucement that  --enable-all-security-services will likely be deprecated and potentially removed later.
I'd like it to be revised since it is helping a lot and solves the most common issue with security services: they are often dynamically configured so static analysis can never find them - and the native binary must not loose that capability.

Can this flag be kept as this?

Romain

**Comment by mcraj017** [comment]
@rmannibucau thanks for reporting the issue
@christianwimmer please advise

**Comment by christianwimmer** [comment]
@rmannibucau do you have an example where the static analysis misses a security service?

Global flags like `--enable-all-security-services` do not compose overly well, because some library adds them to the native-image configuration and then it is enable for all applications that depend on that library. So instead of having flags like this, we need to make sure that the static analysis picks up all the relevant use cases.

**Comment by christianwimmer** [comment]
fyi @cstancu

**Comment by rmannibucau** [comment]
Dont have it handy but in 100% of the apps i saw the service is configurable (because security evolves) by a string so not having an include all (or include all this family if you want to drop others) will hurt.
Look http client and ssl for ex, you dont know upfront what you need and will need so you must include most related ciphered blindly for ex.

**Comment by wirthi** [comment]
The option has already been deprecated with the GraalVM 21 release, as previously announced. I don't think this ticket is relevant any longer.

**Comment by rmannibucau** [comment]
@wirthi please consider this ticket still relevant since there is no real solution. The issue I read is that some library can enable it but if so it is more than fine to forbid it to be enabled from META-INF/native-image and requires it on the args only. When some provider are dynamically registered (or not) based on configuration it is almost impossible to detect it must be registered or not and replacing the auto registration by an explicit one is a lot of work to do, test and maintain compared to the option which was working out of the box.

**Comment by wirthi** [comment]
The flag is unused since March 2021 (!) and the subsequent GraalVM 21.1.0 release. In this and newer GraalVM versions, you can still set it as a deprecated flag, but it has zero effect. We have not had any complaint so far after the deprecation.

Static analysis improved so far that it detects these cases well enough, or it should be easy enough to manually add configuration for it. If you disagree with that (especially the "analysis is good enough" part), please share a reproducer that causes an actual problem in your code base. But most likely, we would advice you to add manual configuration.

**Comment by rmannibucau** [comment]
> We have not had any complaint so far after the deprecation.

Maybe because ticket was already opened ;).

> Static analysis improved so far that it detects these cases well enough, or it should be easy enough to manually add configuration for it.

Can you point me out how it does work and the needed config  - out reflection on the provider itself indeed - to make work this sample (you can take bouncycastle as an example with code doing something like "if (toggle && hasClass(bcprovider)) Security.addProvider(0, load());")

**Comment by wirthi** [comment]
Please share a reproducible application that is not working with native image. Then we can investigate why that is the case and advise you how to solve it. 

A general documentation on the metadata/configuration you can provide [you can find on our website](https://www.graalvm.org/latest/reference-manual/native-image/metadata/). Our customers are using the [Tracing Agent](https://www.graalvm.org/latest/reference-manual/native-image/metadata/AutomaticMetadataCollection/ ) to generate most of their configuration.

**Comment by rmannibucau** [comment]
Ok, checked with mutiples versions of graal (in particular older since I was no more sure of the behavior and I added a lot of configuration to compensate bugs of that time) and you are right, the behavior is still unexpected compared to the JVM but there is no regression disabling the option.

Here are the few cases which are not warmly welcomed in apps:

* `Security.getProviders` ignores sun provider (so when tested before switching or to validate a runtime at startup it fails and needs native-image careness - but does not seem new)
* Bouncycastle using a ton of reflection almost none of the services are found (29 vs 4049 in JVM mode) even if static analyzis enables to find all the related names (reproducer: `Security.addProvider(new BouncyCastleProvider());System.out.println(Security.getProvider("BC").size());`)

Rest looks acceptable for the apps I tested on native-image.

Side note: tested with `22.3.0.r17` (old versioning scheme) and `21.0.2-graalce` (new versioning).
